### ISSUE TITLE
Moving camera causing flicker effect on low-res textures

### ISSUE THREAD

**Issue by meowgoocat** [prime31/Nez#142]
I've been bouncing around quite a bit so I'm not quite sure what the best way to tackle this is.  Basically, the issue is this:

I import a tiled map with a very low-res tileset (roughly 32x32 pixels).  The game is isometric, so the movement is done at an unusual angle (16 X, 13 Y).  Due to this unusual ratio and the low-res art, there is a very distinct blurring/flickering effect when the screen scrolls on a follower camera.

This flickering effect is lowered or nonexistent under the following scenarios:
- Much higher resolution tiles are used (512x512 pixels).  This works even if the same tiles that had the issue are scaled up with point scaling.
- The movement angle is made to be a multiple of 45 degrees. (so cardinal directions, basically)
- I go in and apply a hardcoded scale to every batcher.draw call to increase drawing size

Now, obviously I can fix this by just scaling up my assets, but that's a ridiculous requirement that will significantly increase file size and load times.  Adding the scale to the draw calls introduces all sorts of bugs, particularly with positions and bounding boxes.  I feel like the solution should be in the camera or renderer, since this is entirely a rendering problem.

Is there any way to fix this?

**Comment by prime31** [comment]
This sounds like it could be a case of rounding causing issues. You can turn off auto rounding globally via the static property in the Transform class. Another way to handle rounding would be similar to how TiledMapMover does: keep track of subpixel movement.

**Comment by meowgoocat** [comment]
Yes, I have auto rounding turned off for other reasons, so that isn't the issue.  It seems it's more that the smaller the tile is, the "more" it moves per frame of movement when it is rendered to its new location.  For example, if the border of the tile is 1 pixel wide, the entirety of the border moves when the camera is shifted instead of just parts of it.  However, for some reason, if the scale is drastically increased but the render size is increased by a comparable amount (such that everything appears the same size in both scenarios), then this effect is lessened.  I think it might have something to do with rounding, but since I have Transform.shouldRoundPosition set to false, I don't see how that could be it.

**Comment by meowgoocat** [comment]
I was able to cheese the issue for the most part.  By simply doubling asset resolution and changing IsFixedTimeStep = True (locking to 60 fps, since 144 fps seemed to be making the issue worse), the effect is now reasonably gone.

However, I don't think this is necessarily an ideal solution, since it requires effectively quadrupling asset size and limiting framerate.

**Comment by prime31** [comment]
I don't think what you are seeing here has anything to do with Nez specifically. This just sounds like plain old aliasing artifacts. Any of the standard ways to solve it should work fine such as using linear interpolation on textures for non-pixel perfect, use point for pixel perfect, lock FPS to monitor refresh rate, etc. If you go pixel perfect then you will want to round positions. If not pixel perfect then don't round and let the GPU interpolate.

**Comment by holyxiaoxin** [comment]
@prime31  I'm trying to turn off transform rounding, and realised that `shouldRoundPosition` has been moved to `shouldRoundDestinations` in https://github.com/prime31/Nez/commit/837d9e359885775f6364c9735c37757d3f04328a. I'm not too sure if this solution will work now, if so how do I set `shouldRoundDestinations`? I've tried setting `Graphics.instance.batcher.shouldRoundDestinations = false;`. in Game1 : Core, override `Initialize` method. I can still see my player flickering when i'm using FollowCamera like below:

```cs
var heroEntity = createEntity("player");
heroEntity.addComponent(new PrototypeSprite(16, 16)).setColor(Color.Red);
camera.addComponent(new FollowCamera(heroEntity, camera) { followLerp = 1 }).setCenteredDeadzone(0, 0);
```
